CuPy JIT documentation (#5012)
cupy/cupy: docs/source/tutorial/kernel.rst

.. note::
   Currently, :func:`cupy.fuse` can fuse only simple elementwise and reduction operations.  Most other routines (e.g. :func:`cupy.matmul`, :func:`cupy.reshape`) are not supported.

.. _jit_kernel_definition:

JIT kernel definition
---------------------

The :class:`~cupyx.jit.rawkernel` decorator can define raw CUDA kernel from Python function.

In this section, the Python function wrapped with the decorator is called as *target function*.

The target function consists of the elemental scalar operations, and the users have to manage how parallelize them. CuPy's array operations which automatically parallelize the operation (e.g., :func:`cupy.add`, :func:`cupy.sum`) is not supported. If the custom kernel from the array functions are desired, please refer the :ref:`kernel_fusion` section.

Basic Usage
^^^^^^^^^^^

> :class:`cupyx.jit.rawkernel` decorator can define a CUDA global function.

Reviewer: I feel this line is repetitive of what's been said in the beginning of the section. How about a more descriptive sentence to introduce the example?
```suggestion
Here is a short example for how to write a :class:`cupyx.jit.rawkernel` to copy the values from ``x`` to ``y`` using a grid-stride loop:
```

.. doctest::

   >>> @jit.rawkernel()
   ... def elementwise_copy(x, y, size):
   ...     tid = jit.blockIdx.x * jit.blockDim.x + jit.threadIdx.x
   ...     ntid = jit.gridDim.x * jit.blockDim.x
   ...     for i in range(tid, size, ntid):
   ...         y[i] = x[i]

   >>> size = cupy.uint32(2 ** 22)
   >>> x = cupy.random.normal(size=(size,), dtype=cupy.float32)
   >>> y = cupy.empty((size,), dtype=cupy.float32)

> 

Reviewer: How about adding a check?
```suggestion
   >>> assert (x == y).all()

```

> Two kinds of interfaces  to launch the kenrel are supported. See the document of :class:`cupyx.jit.CudaFunction` for details.

Reviewer: ```suggestion
Two kinds of interfaces to launch the kernel are supported, see the documentation of :class:`cupyx.jit.CudaFunction` for details.
```
btw, is any of the interfaces experimental? If so we probably want to mention it here.

.. doctest::

   >>> elementwise_copy((128,), (1024,), (x, y, size))

.. doctest::

   >>> elementwise_copy[128, 1024](x, y, size)

> Compilation will be deferred until the first function call. CuPy JIT compiler infers the types of arguments at call time.

Reviewer: I think it's worth mentioning the cache:
```suggestion
The compilation will be deferred until the first function call. CuPy's JIT compiler infers the types of arguments at the call time, and will cache the compiled kernels for speeding up any subsequent calls.
```

Typing rule
^^^^^^^^^^^

> The type of local variables are inferred at the first assignment in the function. The first assign instruction must be at the top-level -- in other words, must not be in if/else body or for-loops.

Reviewer: ```suggestion
The types of local variables are inferred at the first assignment in the function. The first assignment must be done at the top-level of the function; in other words, it must *not* be in ``if``/``else`` bodies or ``for``-loops.
```

List of API
^^^^^^^^^^^

* ``cupyx.jit.threadIdx``
* ``cupyx.jit.blockDim``
* ``cupyx.jit.blockIdx``
* ``cupyx.jit.gridDim``
* ``cupyx.jit.syncthreads()``

> * ``cupyx.jit.shared_memory(dtype, size)``

Reviewer: Do we have docstrings for them? Not sure if this would work? e.g.: ``:func:`cupyx.jit.shared_memory```

Limitations
^^^^^^^^^^^

> CuPy's JIT compiler generates CUDA code via Python AST. We decided not to use Python bytecode to analyze the target function to avoid perforamance degradation. The CUDA source code generated from the Python bytecode will not effectively optimized by CUDA compiler, because for-loops and other control statements of the target function are fully transformed to jump instruction when converting the target function to bytecode.

Reviewer: I feel this is an "advantage" of CuPy JIT, not a "limitation"! Maybe we can move this paragraph somewhere earlier, to deliver an early impression that it could be more performant?

Author: Thanks! Fixed in 221c0b5.

> CuPy's JIT compiler uses :py:func:`inspect.getsource` to get the text of the source code of the target function. So the compiler does not work in the following situtaions.

Reviewer: ```suggestion
CuPy's JIT compiler uses :py:func:`inspect.getsource` to get the source code of the target function, so the compiler does not work in the following situations:
```

> * In iPython iterpreter
> * Lambda expression as the target function.

Reviewer: btw how about Jupyter notebooks? I suppose they don't work either (I didn't test)?
```suggestion
* In IPython interpreter
* Lambda expressions as target functions
```

.. note::
   Currently, :func:`cupy.fuse` can fuse only simple elementwise and reduction operations.  Most other routines (e.g. :func:`cupy.matmul`, :func:`cupy.reshape`) are not supported.

.. _jit_kernel_definition:

JIT kernel definition
---------------------

The :class:`cupyx.jit.rawkernel` decorator can create raw CUDA kernels from Python functions.

In this section, a Python function wrapped with the decorator is called a *target function*.

A target function consists of elementary scalar operations, and users have to manage how to parallelize them. CuPy's array operations which automatically parallelize operations (e.g., :func:`~cupy.add`, :func:`~cupy.sum`) are not supported. If a custom kernel based on such array functions is desired, please refer to the :ref:`kernel_fusion` section.

Basic Usage
^^^^^^^^^^^

Here is a short example for how to write a :class:`cupyx.jit.rawkernel` to copy the values from ``x`` to ``y`` using a grid-stride loop:

.. doctest::

>    >>> @jit.rawkernel()

Reviewer: ```suggestion
   >>> @cupyx.jit.rawkernel()
```

   ... def elementwise_copy(x, y, size):
   ...     tid = jit.blockIdx.x * jit.blockDim.x + jit.threadIdx.x
   ...     ntid = jit.gridDim.x * jit.blockDim.x
   ...     for i in range(tid, size, ntid):
   ...         y[i] = x[i]

   >>> size = cupy.uint32(2 ** 22)
   >>> x = cupy.random.normal(size=(size,), dtype=cupy.float32)
   >>> y = cupy.empty((size,), dtype=cupy.float32)

>    >>> assert (x == y).all()

Reviewer: ```suggestion
```

Two kinds of interfaces to launch the kernel are supported, see the documentation of :class:`cupyx.jit.CudaFunction` for details.

.. doctest::

>    >>> elementwise_copy((128,), (1024,), (x, y, size))

Reviewer: ```suggestion
   >>> elementwise_copy((128,), (1024,), (x, y, size))  # RawKernel style
   >>> assert (x == y).all()
```

.. doctest::

>    >>> elementwise_copy[128, 1024](x, y, size)

Reviewer: ```suggestion
   >>> elementwise_copy[128, 1024](x, y, size)  #  Numba style
   >>> assert (x == y).all()
```

.. note::
   Currently, :func:`cupy.fuse` can fuse only simple elementwise and reduction operations.  Most other routines (e.g. :func:`cupy.matmul`, :func:`cupy.reshape`) are not supported.

.. _jit_kernel_definition:

JIT kernel definition
---------------------

The :class:`cupyx.jit.rawkernel` decorator can create raw CUDA kernels from Python functions.

In this section, a Python function wrapped with the decorator is called a *target function*.

A target function consists of elementary scalar operations, and users have to manage how to parallelize them. CuPy's array operations which automatically parallelize operations (e.g., :func:`~cupy.add`, :func:`~cupy.sum`) are not supported. If a custom kernel based on such array functions is desired, please refer to the :ref:`kernel_fusion` section.

Basic Usage
^^^^^^^^^^^

Here is a short example for how to write a :class:`cupyx.jit.rawkernel` to copy the values from ``x`` to ``y`` using a grid-stride loop:

.. doctest::

>    >>> @cupyx.jit.rawkernel()

Reviewer: Need to expose `jit`...😓
```suggestion
   >>> from cupyx import jit
   >>>
   >>> @jit.rawkernel()
```
